pkg/controllers/management/clusterprovisioner/provisioner.go

```diff
@@ -214,6 +231,32 @@ func (p *Provisioner) Updated(cluster *v3.Cluster) (runtime.Object, error) {
 	return obj.(*v3.Cluster), err
 }
 
+func (p *Provisioner) migrateHelmReleases(cluster *v3.Cluster) (*v3.Cluster, error) {
+	if !features.HelmCleanupOnUpgrade.Enabled() ||
+		cluster.Status.Version == nil ||
+		semver.Compare(cluster.Status.Version.GitVersion, "v1.25") < 0 ||
+		apimgmtv3.ClusterConditionHelmReleasesMigrated.IsTrue(cluster) {
+		// Skip cleaning up releases if any of the conditions are true
+		return cluster, nil
+	}
+
+	err := p.cleanupHelmReleases(cluster)
+	if err != nil {
+		return cluster, err
+	}
+
+	// Attempt to manually trigger updating, otherwise it will not be triggered until after exiting reconcile
+	apimgmtv3.ClusterConditionHelmReleasesMigrated.True(cluster)
+	apimgmtv3.ClusterConditionHelmReleasesMigrated.Message(cluster, "Helm releases cleaned up")
+
+	cluster, err = p.Clusters.Update(cluster)
```

Comment:
I would expect the cleanup of all releases in a cluster to take some time. I would also expect that our clusters are objects which see higher-than-usual traffic (i.e. they are often changed). Given both of these facts, I would overall expect you to see a higher-than-usual occurrence of "This object has been changed, please update the latest version and try again" errors.

To fix this, you can do a `cluster, err := p.Clusters.Get(cluster.name)` before updating the condition. 

You can also wrap this in a retry block which, if there's a failure, gets the cluster, updates the condition, and then tries again. This approach can be seen [here](https://github.com/rancher/rancher/blob/1913156a9f32ded561ab100ff0ce7f9135ad8e85/pkg/api/norman/customization/cluster/utils.go#L158).